### PR TITLE
Feature/e2e

### PR DESCRIPTION
**Pull request by Sam-Jeston**
A standalone E2E suite. The entire compose stack must be running to execute this suite.

### CODE REVIEW

File: test/e2e/features/contract_execution.feature [hunk 1]
```diff
@@ -0,0 +1,30 @@
+Feature: Contract Execution
```
Comment (rhyslbw) [inline test/e2e/features/contract_execution.feature]:
```suggestion
Feature: Contract Interaction
```

File: test/e2e/features/contract_execution.feature [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  In order to interact with contracts
```
Comment (rhyslbw) [inline test/e2e/features/contract_execution.feature]:
```
  To interact with initialized Smart Contracts
  As a client of the platform
  I want to call the endpoints to access expressed behaviour
```

File: test/e2e/features/contract_execution.feature [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  As a consumer of the Smart Contract platform
+  I want to execute contract endpoints
+
+  Scenario Outline: contract execution generates a signature requests
```
Comment (rhyslbw) [inline test/e2e/features/contract_execution.feature]:
```suggestion
  Scenario Outline: contract interaction that generates a transaction
```

File: test/e2e/features/contract_execution.feature [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    When I load a contract by address <address>
+    And I subscribe by public key <pk>
+    And I execute against contract <address> with the method <method> and arguments <methodArguments>
```
Comment (rhyslbw) [inline test/e2e/features/contract_execution.feature]:
```suggestion
    And I call the contract <address> with the method <method> and arguments <methodArguments>
```

File: test/e2e/features/contract_execution.feature [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    Then I should receive a signing request against <pk>
```
Comment (rhyslbw) [inline test/e2e/features/contract_execution.feature]:
```suggestion
    Then <pk> should receive a signing request
```

File: test/e2e/features/contract_execution.feature [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+    Examples:
+      | address | method | methodArguments                                                       | pk            |
+      | "abcd"  | "add"  | "{\"number1\": 1, \"number2\": 2, \"originatorPk\": \"myPublicKey\"}" | "myPublicKey" |
+
+  Scenario Outline: a contract can be loaded multiple times without error
```
Comment (rhyslbw) [inline test/e2e/features/contract_execution.feature]: Let's split this into a new feature, renaming to match UL of _initialize_

**Review by rhyslbw — CHANGES_REQUESTED** [review]
Nice work. I'd like us to consider _features_ as the product definition so hoisting `/features` to the root will be a great improvement. The interpretation of Gherkin for the cucumberjs test suite makes sense to be where it is.

It's also a good time to scaffold out our know features, even if they are marked as yet to implement

- contract uninitialization
- providing a list of initialized contracts by address

File: features/contract_unloading.feature [hunk 7]
```diff
@@ -0,0 +1 @@
+# TBI
```
Comment (rhyslbw) [inline features/contract_unloading.feature]: Let's scope the feature and use a `@Todo` tag to exclude it from the test suite

File: features/list_loaded_contracts.feature [hunk 8]
```diff
@@ -0,0 +1,12 @@
+Feature: List Loaded Contracts
+  As a client of the platform
+  I want to be able to list loaded contracts
```
Comment (rhyslbw) [inline features/list_loaded_contracts.feature]:
```suggestion
  I want to be able to list loaded contracts to understand what I can already interact with
```

File: features/contract_loading.feature [hunk 9]
```diff
@@ -0,0 +1,30 @@
+Feature: Contract Loading
+  To interact with contracts on the platform
+  They must first be loaded
```
Comment (rhyslbw) [inline features/contract_loading.feature]:
```suggestion
  As a client of the backend I must first request it be loaded
  So I can then interact with it
```

File: features/contract_loading.feature [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+  However loading is provably an idempotent operation
```
Comment (rhyslbw) [inline features/contract_loading.feature]: I think we can drop this from here
```suggestion
```

File: features/contract_loading.feature [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+
+  Scenario Outline: a contract cannot be called without first being loaded
+    When I subscribe by public key <pk>
+    And I call the contract <address> with the method <method> and arguments <methodArguments> knowing the contract is not initialized
+    Then <pk> should not receive a signing request
```
Comment (rhyslbw) [inline features/contract_loading.feature]: Better to model this outcome as an error. While we could auto-load the contract, I'd suggest we don't as then the two operations remain independent for now

File: features/contract_loading.feature [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+
+    Examples:
+      | address              | pk                | method | methodArguments                       |
+      | "plutusGuessingGame" | "uniquePublicKey" | "add"  | "{\"originatorPk\": \"myPublicKey\"}" |
+
+  Scenario Outline: a contract can be loaded multiple times without error
```
Comment (rhyslbw) [inline features/contract_loading.feature]:
```suggestion
  Scenario Outline: a contract is loaded only once, with subsequent requests ignored
```

**Review by rhyslbw — CHANGES_REQUESTED** [review]
Let's also add in the features `@todo`
**Submitting a signed transaction to the network**
**Client validation** 

We should enforce that in order to be granted a connect, a hash of the expected schema must be passed and verified. This gives us the assurance that the client is using the expected schema without needing to introduce versioning.

Please also add

**Comment by Sam-Jeston** [comment]
I added feature files for the upcoming features, with descriptions. I did not go to the length of declaring scenarios with associated steps. Because of this the use of tags was not required. Happy to get your thoughts on this @rhyslbw

**Review by rhyslbw — APPROVED** [review]

**Pull request merged**